OPAL/si/src/main/scala/org/opalj/fpcf/scheduling/MaximumPhaseScheduling.scala

```diff
@@ -0,0 +1,235 @@
+         * 2. Updates the scheduling graph (`aCyclicGraph`) to reflect changes and remove redundant batches.
+         * 3. Recursively processes the next batch until all necessary batches are updated.
+         * 4. Ensures computations are properly assigned across batches while preserving execution dependencies.
+         */
+        def setLazyInAllBatches(
+            tmp_aCyclicGraph: Map[List[Int], Set[Int]],
+            firstElement:     List[Int]
+        ): Map[List[Int], Set[Int]] = {
+            var visited_batches: List[List[Int]] = List.empty
+            var aCyclicGraph = tmp_aCyclicGraph
+
+            def setLazyInAllBatchesInternal(
+                tmp_aCyclicGraphInternal: Map[List[Int], Set[Int]],
+                firstElement:             List[Int]
+            ): Map[List[Int], Set[Int]] = {
+
+                if (firstElement.forall(csID =>
+                        computationSpecificationArray(csID).computationType.equals(lazyType) ||
+                            computationSpecificationArray(csID).computationType.equals(transformerType)
+                    )
+                ) {
+                    var existInSomeBatch = false
+                    tmp_aCyclicGraphInternal.foreach { batch =>
+                        if (batch._2.toList.intersect(firstElement).nonEmpty && batch._1 != firstElement) {
+                            aCyclicGraph = aCyclicGraph + ((batch._1 ++ firstElement) -> mapCSToNum(
+                                getAllCSFromPropertyBounds(getAllUses(batch._1 ++ firstElement))
+                            ).diff((batch._1 ++ firstElement).toSet))
+                            aCyclicGraph = aCyclicGraph - batch._1
+                            existInSomeBatch = true
+                        }
+                    }
+                    if (existInSomeBatch) {
+                        aCyclicGraph = aCyclicGraph - firstElement
+                        setLazyInAllBatchesInternal(aCyclicGraph, aCyclicGraph.head._1)
+                    } else {
+                        visited_batches = visited_batches :+ firstElement
+                        val keyList = aCyclicGraph.keys.toSet -- visited_batches
+                        if (keyList.nonEmpty) {
+                            aCyclicGraph = setLazyInAllBatchesInternal(aCyclicGraph, keyList.head)
+                        }
+                    }
+                } else {
+                    visited_batches = visited_batches :+ firstElement
+                    val keyList = aCyclicGraph.keys.toSet -- visited_batches
+                    if (keyList.nonEmpty) {
+                        setLazyInAllBatchesInternal(aCyclicGraph, keyList.head)
+                    }
+                }
+                aCyclicGraph
+            }
+            setLazyInAllBatchesInternal(aCyclicGraph, firstElement)
+        }
+
+        def topologicalSort(graph: Map[Int, List[Int]]): List[Int] = {
```

Comment:
This can probably be moved to the package object of `org.opalj.graphs` in OPAL's `Common` subproject.